fix(e2e::network): eliminate the dependency of external ips

knight42 opened the pull request:
Signed-off-by: knight42 <anonymousknight96@gmail.com>


**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation

/kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR refactors two e2e network tests so that they no longer rely on the external IPs of the nodes. 

**Which issue(s) this PR fixes**:

Fixes #93341  

Part of #90047

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

knight42 committed:
fix(e2e): access nodes via test container in LB network tests

Signed-off-by: knight42 <anonymousknight96@gmail.com>

knight42 commented:
/cc @mm4tt @wojtek-t

knight42 commented:
@mm4tt @wojtek-t hi! I need a `/ok-to-test` to test this patch

jprzychodzen commented:
/ok-to-test

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
@mm4tt https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92163/pull-kubernetes-e2e-gce-correctness/1272862058299265025/ the bootstrap script throws an `argparse.ArgumentTypeError` in test `pull-kubernetes-e2e-gce-correctness`

knight42 commented:
@mm4tt @wojtek-t currently the tests cannot be built on CI due to
```
I0616 12:04:15.383] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7977/execroot/io_k8s_kubernetes/test/e2e/network/service.go:806:21: jig.ListNodesWithEndpoint undefined (type *service.TestJig has no field or method ListNodesWithEndpoint)
I0616 12:04:15.384] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7977/execroot/io_k8s_kubernetes/test/e2e/network/service.go:3093:25: config.GetResponseFromTestContainer undefined (type *network.NetworkingTestConfig has no field or method GetResponseFromTestContainer)
I0616 12:04:15.384] /bazel-scratch/.cache/bazel/_bazel_prow/48d5366022b4e3197674c8d6e2bee219/sandbox/processwrapper-sandbox/7977/execroot/io_k8s_kubernetes/test/e2e/network/service.go:3152:21: undefined: GetHTTPContentFromTestContainer
```

is it a cache issue? or should I rebase on top of the current master branch and add back those methods since #91559 was reverted?

jprzychodzen commented:
@knight42, the code you are referring to does not exists anymore - it's not a cache issue. It was reverted.

Please rebase and fix.

jprzychodzen commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
@mm4tt @wojtek-t found a case that external IP may be actually required. The test `should work for type=NodePort` checks the client ip, if we access the service from a test container, we would only get the ip address of the test container rather then the client https://github.com/knight42/kubernetes/blob/001e9fd5bc38ab2e444725e47aa892884efe1467/test/e2e/network/service.go#L3151-L3157

mm4tt commented:
> Matt Matejczyk Wojciech Tyczynski found a case that external IP may be actually required. The test `should work for type=NodePort` checks the client ip, if we access the service from a test container, we would only get the ip address of the test container rather then the client https://github.com/knight42/kubernetes/blob/001e9fd5bc38ab2e444725e47aa892884efe1467/test/e2e/network/service.go#L3151-L3157

Actually, the test pod ip will be the client ip in that case. We should just check that the client ip we got matches the ip of the `TestContainerPod`- https://github.com/kubernetes/kubernetes/blob/708fb6b457d94c770890731e53090a975f53679a/test/e2e/framework/network/utils.go#L123. So the check should be something like:
```
if clientIp != config.TestContainerPod.Status.PodIP {
  framework.Failf("Source IP was NOT preserved")
}
```

knight42 commented:
@mm4tt ah that makes sense, thanks!

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

mm4tt commented on test/e2e/network/service.go:
```diff
@@ -3103,17 +3139,20 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
+			ginkgo.By(fmt.Sprintf("reading clientIP using the TCP service's NodePort, on node %v: %v:%v/%v", nodeName, nodeIP, tcpNodePort, dialCmd))
+			clientIP, err := GetHTTPContentFromTestContainer(config, nodeIP, tcpNodePort, e2eservice.KubeProxyLagTimeout, dialCmd)
+			framework.ExpectNoError(err)
 			framework.Logf("ClientIP detected by target pod using NodePort is %s", clientIP)
```
Can we also log the TestContainerPod ip? It will be easier to debug

knight42 replied:
sure

knight42 commented:
a mysterious phenomenon noticed. According to the lines [123638-123641](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/92163/pull-kubernetes-e2e-gce-correctness/1273461188545482752/#1:build-log.txt%3A123638) from build log, `pull-kubernetes-e2e-gce-correctness` failed because of timeout:

```
I0618 05:01:37.837] {"msg":"PASSED [k8s.io] Pods should cap back-off at MaxContainerBackOff [Slow][NodeConformance]","total":-1,"completed":20,"skipped":49,"failed":0}
I0618 05:01:37.837] Jun 18 05:01:37.829: INFO: Running AfterSuite actions on all nodes
I0618 05:01:37.837] 
W0618 07:45:50.268] 2020/06/18 07:45:50 process.go:199: Interrupt after 4h0m0s timeout during ./hack/ginkgo-e2e.sh --ginkgo.skip=\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]|\[DisabledForLargeClusters\] --minStartupPods=8 --node-schedulable-timeout=90m --logexporter-gcs-path=gs://kubernetes-jenkins/pr-logs/pull/92163/pull-kubernetes-e2e-gce-correctness/1273461188545482752/artifacts --report-dir=/workspace/_artifacts --disable-log-dump=true --cluster-ip-range=10.64.0.0/14. Will terminate in another 15m
```

and the follow-up log shows that the test `[sig-storage] HostPathType Socket [Slow]` stuck somewhere
```
...
I0618 07:45:50.562] [BeforeEach] [sig-storage] HostPathType Socket [Slow]
I0618 07:45:50.562]   test/e2e/framework/framework.go:174
I0618 07:45:50.562] STEP: Creating a kubernetes client
I0618 07:45:50.562] Jun 18 04:31:27.719: INFO: >>> kubeConfig: /workspace/.kube/config
I0618 07:45:50.563] STEP: Building a namespace api object, basename host-path-type-socket
I0618 07:45:50.563] STEP: Waiting for a default service account to be provisioned in namespace
I0618 07:45:50.563] [BeforeEach] [sig-storage] HostPathType Socket [Slow]
I0618 07:45:50.563]   test/e2e/storage/host_path_type.go:190
I0618 07:45:50.563] STEP: Create a pod for further testing
I0618 07:45:50.563] STEP: running on node gce-cluster-minion-group-80gv
I0618 07:45:50.564] STEP: Create a socket for further testing
I0618 07:45:50.565] Jun 18 04:31:30.100: INFO: ExecWithOptions {Command:[/bin/sh -c nc -lU /mnt/test/asocket &] Namespace:host-path-type-socket-9633 PodName:test-hostpath-type-5tcr7 ContainerName:host-path-testing Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:false}
I0618 07:45:50.566] Jun 18 04:31:30.100: INFO: >>> kubeConfig: /workspace/.kube/config
I0618 07:45:50.566] 
I0618 07:45:50.567] ---------------------------------------------------------
I0618 07:45:50.569] Received interrupt.  Running AfterSuite...
I0618 07:45:50.570] ^C again to terminate immediately
I0618 07:45:50.570] Jun 18 07:45:50.279: INFO: Running AfterSuite actions on all nodes
I0618 07:45:50.570] Jun 18 07:45:50.283: INFO: Waiting up to 3m0s for all (but 1) nodes to be ready
I0618 07:45:50.570] STEP: Destroying namespace "host-path-type-socket-9633" for this suite.
...
```

@mm4tt do you have any clues about this? If not, I would like to add more debug logging to the test `HostPathType Socket [Slow]` first.

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
looks like this prblem https://github.com/kubernetes/kubernetes/pull/92163#issuecomment-645893730 has been solved by #92205, will try again.

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/retest

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
@mm4tt @wojtek-t The `pull-kubernetes-e2e-gce-correctness` test finally passed! 🥳

knight42 commented:
/retest

mm4tt commented on test/e2e/network/service.go:
```diff
@@ -3103,42 +3186,72 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
+			var body string
+			pollFn := func() (bool, error) {
+				// we expect connection failure here, but not other errors
+
```
NIT: Remove blank line

mm4tt commented on test/e2e/network/service.go:
```diff
@@ -615,6 +614,56 @@ func testHTTPHealthCheckNodePort(ip string, port int, request string) (bool, err
+		framework.Logf("Got response from health check node port via test container: %#v", rawResponse)
+		var success bool
+		if len(rawResponse.Responses) > 0 {
+			// We cannot get the status code if we curl the healthz endpoint in the test container,
```
Is this really true? Cannot we introduce another method `GetHTTPCodeFromTestContainer` that would be executing something like `curl -w "%{http_code}" -s -o /dev/null <address>`? It'd simplify this method a lot

mm4tt reviewed (commented):
Nice

mm4tt commented:
Also, a suggestion, not sure if applicable to this PR, but definitely a good one for the future: it would be much easier to review if you sent a few smaller PRs instead of one huge :)

knight42 commented:
> it would be much easier to review if you sent a few smaller PRs instead of one huge

It is indeed a good advice and I wish I could split this PR, but I guess this case may be a little more complex.

It turns out that several e2e service tests rely on method `TestJig.GetEndpointNodes()`, but it is implemented incorrectly and would return empty map when nodes have no external IP, which makes the test almost no-op. If we fix the `GetEndpointNodes` method, we would break these tests because we cannot access nodes without external IP, so I have to fix them at the same time. I think we are just repaying unknown technical debt.

knight42 commented:
with this patch, I think the following two functions can be safely removed as they are not used any more and they are not referenced in other packages.

https://github.com/kubernetes/kubernetes/blob/22f71716cd78cc0c4db70c3517125921269aee7e/test/e2e/network/service.go#L568

https://github.com/kubernetes/kubernetes/blob/22f71716cd78cc0c4db70c3517125921269aee7e/test/e2e/network/service.go#L588

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

mm4tt commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

mm4tt commented:
> with this patch, I think the following two functions can be safely removed as they are not used any more and not referenced in other packages.
> 
> https://github.com/kubernetes/kubernetes/blob/22f71716cd78cc0c4db70c3517125921269aee7e/test/e2e/network/service.go#L568
> 
> https://github.com/kubernetes/kubernetes/blob/22f71716cd78cc0c4db70c3517125921269aee7e/test/e2e/network/service.go#L588

In that case let's remove them (can be in a separate PR). Thanks!

knight42 commented:
got it! I would open another PR to remove these functions once this is merged.

mm4tt commented on test/e2e/network/service.go:
```diff
@@ -3103,42 +3163,71 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
 			framework.Failf("Service HealthCheck NodePort still present")
 		}
 
-		endpointNodeMap, err := jig.GetEndpointNodes()
```
Given that we have a presubmit now, let's remove the `DisabledForLargeClusters` tag from this particular test to make sure our fix works.

mm4tt reviewed (commented):
Looks really good, please squash all the commits and I think we'll be ready to submit

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/retest

knight42 commented:
/test pull-kubernetes-e2e-kind-ipv6

knight42 commented:
@mm4tt I have removed the tag as you said and the test passed, looks like our fix actually works 🎉

and you might want to take a second look at this commit https://github.com/kubernetes/kubernetes/pull/92163/commits/f1f18aa317e1cf284d34be2e1f6f83046ed3e757, I have changed the method `GetHTTPCodeFromTestContainer` a bit. If everything is ok, then I would squash all the commits.

mm4tt commented:
> GetHTTPCodeFromTestContainer

Looks great, let's squash and submit

knight42 commented:
/test pull-kubernetes-e2e-gce-correctness

knight42 commented:
/retest

knight42 commented:
/test pull-kubernetes-integration

knight42 commented:
@mm4tt all tests passed!

mm4tt commented:
Nice!

/lgtm

@wojtek-t for final approval

wojtek-t commented on test/e2e/framework/network/utils.go:
```diff
@@ -334,6 +330,50 @@ func (config *NetworkingTestConfig) GetEndpointsFromContainer(protocol, containe
+	// We only care about the status code reported by curl,
+	// and want to return any other errors, such as cannot execute command in the Pod.
+	// If curl failed to connect to host, it would exit with code 7, which makes `ExecShellInPodWithFullOutput`
+	// return a non-nil error and output "000" to stdout.
```
000 ?

Then this method would return 0, which is not what we want, right?

knight42 replied:
it is kind of difficult to tell the errors caused by curl(such as fail to connect to remote, resolve host, etc.) from the errors caused by executing command in the test container(such as fail to connect to the api server). IMHO the former errors are retriable while the latter are not(I suppose it indicates that the cluster might be misconfigured), I think returning 0 is sufficient to let users know curl is sucessfully executed but it cannot connect to the host.

I am not sure if there is a better choice, any suggestion is welcome.

wojtek-t commented on test/e2e/network/service.go:
```diff
@@ -683,6 +708,23 @@ func waitForApiserverUp(c clientset.Interface) error {
 
+// getEndpointNodesWithInternalIP returns a map of nodenames:internal-ip on which the
+// endpoints of the Service are running.
+func getEndpointNodesWithInternalIP(jig *e2eservice.TestJig) (map[string]string, error) {
```
Let's move this function to jig.go (make it method of jig the same way as GetEndpointNodes is.

knight42 replied:
Currently we already have a similar method, i.e. `jig.GetEndpointNodesWithIP(corev1.NodeInernalIP)`，it returns a map of node names -> list of internal IPs and would return empty list if no IP found.

I put the function `getEndpointNodesWithInternalIP` here because I think this function would only be used here, and it returns a map of node names -> ip(rather than a list of IP) and it would return an error if no IP found.

wojtek-t reviewed (commented):
@knight42 - will you be able to move changes to both util files and jig to a separate PR. I'm fairly confident about those and we can proceed with them, whereas the test changes themselves I would like someone from the networking team to also take a look.

knight42 commented:
@wojtek-t I would like to split this PR, but as I said in https://github.com/kubernetes/kubernetes/pull/92163#issuecomment-647630762, since I have fixed the method `TestJig.GetEndpointNodes()`, if the tests were not updated we would break these tests.

knight42 commented:
@wojtek-t friendly ping. what should I do next?

mm4tt commented:
Hey @knight42, Wojtek will be OOO for the next 2 weeks. I'm looping in network folks to get approval from them.

/assign @bowei 
/assign @freehan

thockin reviewed (commented):
@wojtek-t Do you have any lingering concerns?  This looks like a strict improvement to me, and I am OK with the remaining comments' answers.

/approve
/lgtm
/hold

Hold in case @wojtek-t wants to say more, but feel free to remove this in a few days.  We can always iterate

wojtek-t commented:
Yeah - I'm fine with the answers too.
Given it's actually significant improvement, I also want this to be in asap.

/hold cancel

Given we're still before test deadline, can we add milestone 1.19 to it? @hasheddan @hkamel ?

wojtek-t commented:
Thanks @thockin  for approving it BTW - I didn't feel competent for this one...

hasheddan commented:
/milestone v1.19

wojtek-t commented:
/retest
